packages/flutter/lib/src/material/button_style.dart

```diff
@@ -0,0 +1,513 @@
+    MaterialStateProperty<double> elevation,
+    MaterialStateProperty<EdgeInsetsGeometry> padding,
+    MaterialStateProperty<Size> minimumSize,
+    MaterialStateProperty<BorderSide> side,
+    MaterialStateProperty<OutlinedBorder> shape,
+    MaterialStateProperty<MouseCursor> mouseCursor,
+    VisualDensity visualDensity,
+    MaterialTapTargetSize tapTargetSize,
+    Duration animationDuration,
+    bool enableFeedback,
+  }) {
+    return ButtonStyle(
+      textStyle: textStyle ?? this.textStyle,
+      backgroundColor: backgroundColor ?? this.backgroundColor,
+      foregroundColor: foregroundColor ?? this.foregroundColor,
+      overlayColor: overlayColor ?? this.overlayColor,
+      shadowColor: shadowColor ?? this.shadowColor,
+      elevation: elevation ?? this.elevation,
+      padding: padding ?? this.padding,
+      minimumSize: minimumSize ?? this.minimumSize,
+      side: side ?? this.side,
+      shape: shape ?? this.shape,
+      mouseCursor: mouseCursor ?? this.mouseCursor,
+      visualDensity: visualDensity ?? this.visualDensity,
+      tapTargetSize: tapTargetSize ?? this.tapTargetSize,
+      animationDuration: animationDuration ?? this.animationDuration,
+      enableFeedback: enableFeedback ?? this.enableFeedback,
+    );
+  }
+
+  /// Returns a copy of this ButtonStyle where the non-null fields in [style]
+  /// have replaced the corresponding null fields in this ButtonStyle.
+  ///
+  /// In other words, [style] is used to fill in unspecified (null) fields
+  /// this ButtonStyle.
+  ButtonStyle merge(ButtonStyle style) {
+    if (style == null)
+      return this;
+    return copyWith(
+      textStyle: textStyle ?? style.textStyle,
+      backgroundColor: backgroundColor ?? style.backgroundColor,
+      foregroundColor: foregroundColor ?? style.foregroundColor,
+      overlayColor: overlayColor ?? style.overlayColor,
+      shadowColor: shadowColor ?? style.shadowColor,
+      elevation: elevation ?? style.elevation,
+      padding: padding ?? style.padding,
+      minimumSize: minimumSize ?? style.minimumSize,
+      side: side ?? style.side,
+      shape: shape ?? style.shape,
+      mouseCursor: mouseCursor ?? style.mouseCursor,
+      visualDensity: visualDensity ?? style.visualDensity,
+      tapTargetSize: tapTargetSize ?? style.tapTargetSize,
+      animationDuration: animationDuration ?? style.animationDuration,
+      enableFeedback: enableFeedback ?? style.enableFeedback,
+    );
+  }
+
+  @override
+  int get hashCode {
+    return hashValues(
+      textStyle,
+      backgroundColor,
+      foregroundColor,
+      overlayColor,
+      shadowColor,
+      elevation,
+      padding,
+      minimumSize,
+      side,
+      shape,
+      mouseCursor,
+      visualDensity,
+      tapTargetSize,
+      animationDuration,
+      enableFeedback,
+    );
+  }
+
+  @override
+  bool operator ==(Object other) {
+    if (identical(this, other))
+      return true;
+    if (other.runtimeType != runtimeType)
```

Comment:
I think the two tests mean something slightly different. This test proves that `this` and `other` have the exact same type.  The `other is ButtonStyle` is weaker, since other could be a ButtonStyle subclass.